strutil/set.go

```diff
@@ -0,0 +1,85 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package strutil
+
+// OrderedSet is a set of strings that maintains the order of insertion.
+//
+// The order of putting items into the set is retained. Putting items "a", "b"
+// and then "a", results in order "a", "b".
+//
+// External synchronization is required for safe concurrent access.
+type OrderedSet struct {
+	positionOf map[string]int
+}
+
+// Items returns a slice of strings representing insertion order.
+//
+// Items is O(N) in the size of the set.
+func (o *OrderedSet) Items() []string {
+	if len(o.positionOf) == 0 {
+		return nil
+	}
+	items := make([]string, len(o.positionOf))
+	for item, idx := range o.positionOf {
+		items[idx] = item
+	}
+	return items
+}
+
+// Contains returns true if the set contains a given item.
+//
+// Contains is O(1) in the size of the set.
+func (o *OrderedSet) Contains(item string) bool {
+	if len(o.positionOf) == 0 {
```

Comment:
Do we really need this check? It seems like we could just do:
```
       _, ok := o.positionOf[item]
	return ok
```
I don't think we actually safe much instructions with the extra if() (but I'm happy to be proven wrong).

Reply:
I wanted to avoid working with nil maps.

Reply:
Though on second thought this should work as well.

Reply:
Though on second thought this should work as well.